### ISSUE TITLE
NotificationBroadcastReceiver not receiving notifications; ViewPager

### ISSUE THREAD

**Issue by aromanelli** [google/ExoPlayer#4643]
### Issue description
TL;DR Activity with a ViewPager, with ten-ish Fragments, most, but not all of them, each having an ExoPlayer, the later fragments will not have their notification play buttons work, while the first/earlier ones will.  Notification buttons always break after there's a gap in the ViewPager display hierarchy, of a fragment without an ExoPlayer in it.

Long Version:  I have an activity, which uses a ViewPager, which instantiates (via the ViewPager adapter) fragments, ten-ish.  Many fragment has an SimpleExoPlayer in it, but a few fragments do not (they just have an ImageView).  Each player gets set to a PlayerNotificationManager.  Each PlayerNotificationManager has a class implementer of MediaDescriptionAdapter set to it.

Depending on the layout/orientation, a fragment may get instantiated early, and held in memory, or instantiated real time, as the user swipes back/forth.

For the first fragment, showing a SimpleExoPlayer, and having a notification, the notification Play/Back buttons work fine (com.google.android.exoplayer.play).  The NotificationBroadcastReceiver, inside of the PlayerNotificationManager, has its onReceive(Context, Intent) called, when the buttons on the notification are pressed.

If swiping to the right once, you would have a non-SimpleExoPlayer fragment.  Swiping once more to the right, you would have a fragment with another SimpleExoPlayer, and a notification shows for it.  However, when pressing the buttons on the notification, nothing happens.  The broadcast receiver's onReceive method is not called.

If I restart the app, and go to a later page/fragment, the notification buttons may or may not work.  If I manually slide back and forth some, eventually, I can get some page's/fragment's notification buttons to work, even if they were not working previously in the same app run.

If I change the adapter to only instantiate one of the later fragments, in the view pager, it'll work fine.

The common thing that tends to break the notification buttons is as soon as you go past one/two view pages of a page/fragment that does not have a video player in it.  Some kind of internal tracking mechanism that assumes all pages in a ViewPager would have a player, maybe?

### Reproduction steps
Use a Nexus 10, API 26 (or any tablet)(though you can also see this via a phone/portrait, Nexus 6).  The code/app you will find in my github repo, link below.

Be sure to set a breakpoint in PlayerNotificationManager.NotificationBroadcastReceiver.onReceive(Context, Intent) beforehand.

Run the app.  Select "Brownies".  On the left side there is an ingredient, then a list of steps.  Select the first step, which is the second entry from the top, "Step: Recipe Introduction".  A player will display, and a notification icon shows at the top.  Pull down the notification drawer, and hit the play button.  The video will play.

Then, swipe from right to left.  You'll see a new fragment/page with only an image in it, no video player.  Note that there is no notification at the top.  Swipe right to left again.

You'll be at another fragment/page, and another notification icon will display.  Pull down the notification drawer.  The video may or may not play.  If it plays, swipe right to left, passing by any image pages, until you get to a page that has a video player that the notification play button does not work on.  It'll happen by the time you get 1/2 the way through the pages/fragments (there' usually 9-10ish), usually more no than one/two more swipes.

If you play around with the app, by swiping away several pages, then coming back, you can sometimes get a page/fragment's notification to work, that did not work, even during the same run.

Finally, the recycler view on the left, you can direct click on any of the entries, to take you straight to that page/fragment.  If you click on the first step, second entry in recycler view, it always works.  If you start the app, and click right away, without first swiping, one of the later pages/fragments, it won't work.  Note that the ViewPager adapter will instantiate pages/fragments to the left/right of the target page/fragement.

### Link to test content
https://github.com/aromanelli/BakingApp/blob/exoplayerviewpagernotify/app/src/main/java/info/romanelli/udacity/bakingapp/RecipeInfoStepFragment.java

### Version of ExoPlayer being used
'com.google.android.exoplayer:exoplayer:2.8.2'

### Device(s) and version(s) of Android being used
Emulator only, Nexus 6 and 10, API 26.

Final note.  I spent some time trying to debug PlayerNotificationManager/NotificationBroadcastReceiver, but could not find the problem.  I'm fairly new to Android programming, though a long-time JavaFX/Java Swing rich-client coder, so I'm properly dangerous.  I can't imagine, if its just operator error, what I did wrong, even though its complicated code.  I was dealing with issues of the notification manager putting notification icons in the notification drawer for fragments/players that were not visible yet, or not showing notifications for pages/fragments that were visible, but had been already instantiated by the ViewPager's adapter, etc., so I may be doing something not according to spec, but it seems very straightforward that some pages get notifications with buttons that work, and some don't, and if I change how many fragments I load, pages/fragments that would not work before, work now.

**Comment by marcbaechinger** [comment]
I think the problem is around having two instances of PlayerNotificationManager registering a broadcast receiver. You should try using a single PlayerNotificationManager only.

I'd also suggest to have only a single player instance for all fragments in the ViewPager. 

That's a similar arrangement like for a RecyclerView. If you are using more than one instance of the player, you need to make sure that a given device has enough codec instances. This is not trivial because there's no reliable API to check the number of available codec instances.

If you choose to use more than a single ExoPlayer instance I'd still use only one instance of the PlayerNotificationManager and then use playerNotificationManager.setPlayer(player) with the player instance which is currently playing and for which you want to have a notification displayed.

But to make things easy, I'd go with one single player instance and one single PlayerNotificationManager instance and reuse these in the fragments. This also saves you from the problem to make sure that you've never two notifications displayed which may happen when you have two player instance/two notification manager at once.

**Comment by aromanelli** [comment]
I'll do a refactor and let you know how it turns out.

**Comment by ojw28** [comment]
@marcbaechinger - I think all those suggestions are good, and have other important benefits. But as a separate point, I'm not sure I understand why this doesn't work correctly, and I wonder whether we could fix it regardless?

Intuitively I think I'd expect all registered receivers to get a broadcast triggered from any notification. Which would not result in correct behavior, but could possibly be fixed by putting an extra identifier into the intents, or something along those lines. That said, what I think I'd expect doesn't sound consistent with the described behavior. So... What does actually happen? And do you think it's fixable as a task that's independent from whether designing an app in this way is a good idea or not :)?

**Comment by aromanelli** [comment]
@marcbaechinger Regardless of how good/bad my beginners design was, please be mindful of the fact that ViewPager instantiates multiple fragments simultaneously, even though its only just showing one fragment on the screen/app at any one time.  

So having just one player/notifier manager, while trying to make sure I'm only showing the video/notifications for the one visible fragment, may be problematic to coordinate, especially if you want to show each fragment-specific text as part of the notification. My hope was that the manager would do this for me. 

Again, with my new/limited understanding of Android development design.  Please feel free to comment on any bad design that may be cause of the bug, I won't mind.

**Comment by marcbaechinger** [comment]
@ojw28 I'll create a test fixture with multiple PlayerNotificationManager's to see what happens.

**Comment by marcbaechinger** [comment]
I'm not sure what actually happens in the application of @aromanelli. However, in my test ficture I found the following behaviour:

I created two sets of a SimpleExoPlayer, a PlayerView and a PlayerNotificationManager with the same notification channel but a different NOTIFICATION_ID. This results in two notifications are shown in the notification drawer. When using the buttons in the notification, only the buttons of the second notification send an action, but these are sent to both broadcast receivers.

To fix that behaviour, we need to prefix (or suffix) the names of the actions for each player notification to make sure the actions are only delivered to the broadcast receiver which belongs to the given PlayerNotificationManager. This prefixing must be applied to all actions and also the custom actions of the CustomActionProvider to have everything kept separated nicely.

**Comment by aromanelli** [comment]
@marcbaechinger Concerning your last comment.  You've hit upon the issue I'm having, where the buttons in the notification only work for earlier instantiated/displayed fragments, and not later.  You could do fast flipping through fragments/pages, and get a fragment whose notification buttons would not work before, to start working, sometimes.

Also, please be aware, that if you use different NOTIFICATION_ID values, you'll have multiple notifications icons displayed, one for each fragment, even though only one fragment is visible, and the others are hidden (instantiated by ViewPager, but not displayed yet).  I would expect that only one notification icon/message would display, for the one that is actually visible in the ViewPager.

**Comment by ojw28** [comment]
> Also, please be aware, that if you use different NOTIFICATION_ID values, you'll have multiple notifications icons displayed, one for each fragment, even though only one fragment is visible, and the others are hidden (instantiated by ViewPager, but not displayed yet). I would expect that only one notification icon/message would display, for the one that is actually visible in the ViewPager.

This doesn't sound like the responsibility of the ExoPlayer library. I don't think we have any way of telling what's visible or not in your application UI. Whereas your application code does, and so ultimately is responsible for managing any necessary lifecycles accordingly.

**Comment by aromanelli** [comment]
> Also, please be aware, that if you use different NOTIFICATION_ID values, you'll have multiple notifications icons displayed, one for each fragment, even though only one fragment is visible, and the others are hidden (instantiated by ViewPager, but not displayed yet). I would expect that only one notification icon/message would display, for the one that is actually visible in the ViewPager.

> This doesn't sound like the responsibility of the ExoPlayer library. I don't think we have any way of telling what's visible or not in your application UI. Whereas your application code does, and so ultimately is responsible for managing any necessary lifecycles accordingly.

Well, I was making that assumption because the way the ViewPager/adapter works is pre-instantiating fragments/pages before they are visible, so if I want to have each fragment have its own notification id/icon, it doesn't seem like its possible.  I was hoping that a simple "is visible?" check inside the library could handle if the notification should be displayed or not.  I'm coming from JavaFX/Java-Swing, where its easy to do a 'isVisible()' type of call, so I'm assuming Android has that same kind of functionality.

If you look at my code that I submitted for this bug, I tried doing manually managing the notifications so only the one that should be shown, is shown, and its allot of code that seems brittle/problematic (though to be fair, I am new to Android programming).  I ended up going with a single notification id just for the reason you mentioned, to help cut down on the 'management' coding needed.

Otherwise, you put allot of the responsibility on the developer to manage the state of the views, when the library is already managing the notifications on its own, and when a simple visibility check in the library would seem to be able to handle the same situation more cleanly, IMHO.  I would urge a solution along those lines, if possible.  If Android lacks that api (blows my mind if it does lack), then I'd understand not implementing that.

**Comment by ojw28** [comment]
I think you're assuming components are more tightly coupled than they are, and making assumptions about how notifications should work that don't necessarily apply to other use cases. ExoPlayer as a library knows nothing about your ViewPager/Adapter. There's no real requirement that the player is bound to any view in the view hierarchy for a notification to be shown by `PlayerNotificationManager`, and even if there were, in many cases it wouldn't make sense to link the notification's visibility to the visibility of a corresponding view. Conversely, one of the primary uses of a playback notification would explicitly be to still allow control of the player when it's _not_ visible elsewhere in the UI.

It's your application code's responsibility to define the behavior you want and link things together. It should be reasonably easy for you to work out which views are visible in your app's UI, even within a ViewPager (I found a bunch of related StackOverflow questions and blog posts with a quick search). You do of course need to write some code to do this, and put some logic in place to achieve whatever behavior's you want, but that's kind of necessary given different apps will want different behaviors.
